java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java

```diff
@@ -114,14 +116,14 @@ private static Type targetType(TypeCastTree typeCastTree) {
         return null;
       case VARIABLE:
         return ((VariableTree) parent).symbol().type();
-      case METHOD_INVOCATION:
-        // FIXME(SONARJAVA-3191) dead code
-        // If placed in invocation arguments, parent of typecast will be ARGUMENTS, not METHOD_INVOCATION
-        MethodInvocationTree mit = (MethodInvocationTree) parent;
-        if (mit.symbol().isMethodSymbol()) {
-          Symbol.MethodSymbol sym = (Symbol.MethodSymbol) mit.symbol();
-          int castArgIndex = mit.arguments().indexOf(typeCastTree);
-          return sym.parameterTypes().get(castArgIndex);
+      case ARGUMENTS:
+        Arguments arguments = (Arguments) parent;
+        if (arguments.parent().is(Tree.Kind.METHOD_INVOCATION)) {
+          MethodInvocationTree mit = (MethodInvocationTree) arguments.parent();
+          return targetTypeFromMethodSymbol(mit.symbol(), mit.arguments(), typeCastTree);
+        } else if (arguments.parent().is(Tree.Kind.NEW_CLASS)) {
+          NewClassTree nct = (NewClassTree) arguments.parent();
+          return targetTypeFromMethodSymbol(nct.constructorSymbol(), nct.arguments(), typeCastTree);
         }
```

Comment:
It can be even more simplified in order to avoid calling 4 times `arguments.parent()`, and reusing `arguments`. It also makes code a bit easier to read:
```java
Arguments arguments = (Arguments) parent;
Tree invocation = arguments.parent();
if (invocation.is(Tree.Kind.METHOD_INVOCATION)) {
  return targetTypeFromMethodSymbol(((MethodInvocationTree) invocation).symbol(), arguments, typeCastTree);
}
// no need for else, early return
if (invocation.is(Tree.Kind.NEW_CLASS)) {
  return targetTypeFromMethodSymbol(((NewClassTree) invocation).constructorSymbol(), arguments, typeCastTree);
}
return null;
```
In term of code coverage, the only 3rd case of Arguments tree being used is when using annotations (see `AnnotationTree`). It worth adding a test case (if not already present). Note that for me it's OK if we don't consider these cases, as annotations parameters are always constant expression... I don't see many cases where wrong/redundant type cast would occur

```java
public class A {

  @MyAnnotation((int) (0L + 42))
             // ^^^^^^^^^^^^^^^ type cast in arguments from AnnotationTree
  Object field;

  @interface MyAnnotation {
    int value() default 42;
  }
}
```